File: throttle_example/src/hashseq.ts
namespace HashSeq {
    interface IMask {
        m0: number;
        m1: number;
    }

    export interface ISolution {
        s0: number;
        s1: number;
    }

    function permute(state: Uint32Array): void {
        state[0] += state[4]; state[12] = (((state[12] ^ state[0]) >>> 16) | ((state[12] ^ state[0]) << (32 - 16))); state[8] += state[12]; state[4] = (((state[4] ^ state[8]) >>> 12) | ((state[4] ^ state[8]) << (32 - 12))); state[0] += state[4]; state[12] = (((state[12] ^ state[0]) >>> 8) | ((state[12] ^ state[0]) << (32 - 8))); state[8] += state[12]; state[4] = (((state[4] ^ state[8]) >>> 7) | ((state[4] ^ state[8]) << (32 - 7)));
        state[1] += state[5]; state[13] = (((state[13] ^ state[1]) >>> 16) | ((state[13] ^ state[1]) << (32 - 16))); state[9] += state[13]; state[5] = (((state[5] ^ state[9]) >>> 12) | ((state[5] ^ state[9]) << (32 - 12))); state[1] += state[5]; state[13] = (((state[13] ^ state[1]) >>> 8) | ((state[13] ^ state[1]) << (32 - 8))); state[9] += state[13]; state[5] = (((state[5] ^ state[9]) >>> 7) | ((state[5] ^ state[9]) << (32 - 7)));
        state[2] += state[6]; state[14] = (((state[14] ^ state[2]) >>> 16) | ((state[14] ^ state[2]) << (32 - 16))); state[10] += state[14]; state[6] = (((state[6] ^ state[10]) >>> 12) | ((state[6] ^ state[10]) << (32 - 12))); state[2] += state[6]; state[14] = (((state[14] ^ state[2]) >>> 8) | ((state[14] ^ state[2]) << (32 - 8))); state[10] += state[14]; state[6] = (((state[6] ^ state[10]) >>> 7) | ((state[6] ^ state[10]) << (32 - 7)));
        state[3] += state[7]; state[15] = (((state[15] ^ state[3]) >>> 16) | ((state[15] ^ state[3]) << (32 - 16))); state[11] += state[15]; state[7] = (((state[7] ^ state[11]) >>> 12) | ((state[7] ^ state[11]) << (32 - 12))); state[3] += state[7]; state[15] = (((state[15] ^ state[3]) >>> 8) | ((state[15] ^ state[3]) << (32 - 8))); state[11] += state[15]; state[7] = (((state[7] ^ state[11]) >>> 7) | ((state[7] ^ state[11]) << (32 - 7)));

        state[0] += state[5]; state[15] = (((state[15] ^ state[0]) >>> 16) | ((state[15] ^ state[0]) << (32 - 16))); state[10] += state[15]; state[5] = (((state[5] ^ state[10]) >>> 12) | ((state[5] ^ state[10]) << (32 - 12))); state[0] += state[5]; state[15] = (((state[15] ^ state[0]) >>> 8) | ((state[15] ^ state[0]) << (32 - 8))); state[10] += state[15]; state[5] = (((state[5] ^ state[10]) >>> 7) | ((state[5] ^ state[10]) << (32 - 7)));
        state[1] += state[6]; state[12] = (((state[12] ^ state[1]) >>> 16) | ((state[12] ^ state[1]) << (32 - 16))); state[11] += state[12]; state[6] = (((state[6] ^ state[11]) >>> 12) | ((state[6] ^ state[11]) << (32 - 12))); state[1] += state[6]; state[12] = (((state[12] ^ state[1]) >>> 8) | ((state[12] ^ state[1]) << (32 - 8))); state[11] += state[12]; state[6] = (((state[6] ^ state[11]) >>> 7) | ((state[6] ^ state[11]) << (32 - 7)));
        state[2] += state[7]; state[13] = (((state[13] ^ state[2]) >>> 16) | ((state[13] ^ state[2]) << (32 - 16))); state[8] += state[13]; state[7] = (((state[7] ^ state[8]) >>> 12) | ((state[7] ^ state[8]) << (32 - 12))); state[2] += state[7]; state[13] = (((state[13] ^ state[2]) >>> 8) | ((state[13] ^ state[2]) << (32 - 8))); state[8] += state[13]; state[7] = (((state[7] ^ state[8]) >>> 7) | ((state[7] ^ state[8]) << (32 - 7)));
        state[3] += state[4]; state[14] = (((state[14] ^ state[3]) >>> 16) | ((state[14] ^ state[3]) << (32 - 16))); state[9] += state[14]; state[4] = (((state[4] ^ state[9]) >>> 12) | ((state[4] ^ state[9]) << (32 - 12))); state[3] += state[4]; state[14] = (((state[14] ^ state[3]) >>> 8) | ((state[14] ^ state[3]) << (32 - 8))); state[9] += state[14]; state[4] = (((state[4] ^ state[9]) >>> 7) | ((state[4] ^ state[9]) << (32 - 7)));
    }

    function hash_init(state: Uint32Array, suffix: Uint32Array, level: number, iteration: number): void {
        state[0] = 0x6a09e667; state[1] = 0xbb67ae85; state[2] = 0x3c6ef372; state[3] = 0xa54ff53a;
        state[4] = 0x510e527f; state[5] = 0x9b05688c; state[6] = 0x1f83d9ab; state[7] = 0x5be0cd19;

        state[8] = state[0] ^ suffix[0];
        state[9] = state[1] ^ suffix[1];
        state[10] = state[2] ^ suffix[2];
        state[11] = state[3] ^ suffix[3];
        state[12] = state[4] ^ suffix[4];
        state[13] = state[5] ^ suffix[5];
        state[14] = state[6] ^ suffix[6];
        state[15] = state[7] ^ suffix[7];

        state[7] ^= ((level << 16) | iteration);
    }

    function hash_try(ostate: Uint32Array, istate: Uint32Array, proposal: ISolution, mask: IMask): boolean {
        istate[0] = 0x6a09e667 ^ proposal.s0;
        istate[1] = 0xbb67ae85 ^ proposal.s1;
        ostate.set(istate);
        for (let i = 0; i < 6; i++) {
            permute(ostate);
        }
        let f0 = ostate[0];
        let f1 = ostate[1];
        for (let i = 2; i < 16; i += 2) {
            f0 ^= ostate[i];
            f1 ^= ostate[i + 1];
        }
        return ((f0 & mask.m0) | (f1 & mask.m1)) === 0;
    }

    function mask_from_level(mask: IMask, level: number): void {
        if (level > 32) {
            mask.m0 = ~0;
            mask.m1 = (1 << (level - 32)) - 1 | 0;
        } else {
            mask.m1 = 0;
            mask.m0 = (1 << level) - 1 | 0;
        }
    }

    function solve1(suffix: Uint32Array, level: number, iteration: number): ISolution {
        const istate = new Uint32Array(16);
        const ostate = new Uint32Array(16);
        const mask = { m0: 0, m1: 0 };
        hash_init(istate, suffix, level, iteration);
        mask_from_level(mask, level);
        const proposal = { s0: 0, s1: 0 };
        while (!hash_try(ostate, istate, proposal, mask)) {
            proposal.s0 = (proposal.s0 + 1) | 0;
            if (proposal.s0 === 0) {
                proposal.s1 = (proposal.s1 + 1) | 0;
            }
        }
        suffix.set(ostate.slice(8));
        return proposal;
    }

    function verify1(proposal: ISolution, suffix: Uint32Array, level: number, iteration: number): boolean {
        const state = new Uint32Array(16);
        const mask = { m0: 0, m1: 0 };
        hash_init(state, suffix, level, iteration);
        mask_from_level(mask, level);
        if (hash_try(state, state, proposal, mask)) {
            suffix.set(state.slice(8));
            return true;
        }
        return false;
    }

    export function solve(suffix: Uint32Array, levelFirst: number, levelLast: number, iterations: number): ISolution[] {
        const suffix2 = suffix.slice();
        const solutions = [];
        for (let level = levelFirst; level <= levelLast; level++) {
            for (let iteration = 0; iteration < iterations; iteration++) {
                solutions.push(solve1(suffix2, level, iteration));
            }
        }
        return solutions;
    }

    export function solveAsync(callback: (solutions: ISolution[]) => void, suffix: Uint32Array, levelFirst: number, levelLast: number, iterations: number): void {
        const suffix2 = suffix.slice();
        const solutions: ISolution[] = [];

        function solve1Async(level: number, iteration: number): void {
            solutions.push(solve1(suffix2, level, iteration));
            if (++iteration >= iterations) {
                iteration = 0;
                if (++level > levelLast) {
                    return callback(solutions);
                }
            }
            setTimeout(() => { solve1Async(level, iteration); }, 0);
        }
        solve1Async(levelFirst, 0);
    }

    export function verify(solutions: ISolution[], suffix: Uint32Array, levelFirst: number, levelLast: number, iterations: number): boolean {
        const suffix2 = suffix.slice();
        let i = 0;
        for (let level = levelFirst; level <= levelLast; level++) {
            for (let iteration = 0; iteration < iterations; iteration++) {
                if (!verify1(solutions[i++], suffix2, level, iteration)) {
                    return false;
                }
            }
        }
        return true;
    }
}

File: throttle_example/src/hashseq.c

/*
 * A simple hashcash-like proof-of-work, using the BLAKE2S internal permutation.
 *
 * This implementation uses only 32-bit registers in order to be
 * efficient on constrained devices and to be straightforward to
 * port to Javascript.
 */

#include <stdint.h>
#include <stdlib.h>
#include <string.h>

#include "common.h"
#include "hashseq.h"

typedef struct Mask_ {
    uint32_t m0;
    uint32_t m1;
} Mask;

static inline void permute(uint32_t state[16])
{
    G(state[0], state[4], state[8], state[12]);
    G(state[1], state[5], state[9], state[13]);
    G(state[2], state[6], state[10], state[14]);
    G(state[3], state[7], state[11], state[15]);

    G(state[0], state[5], state[10], state[15]);
    G(state[1], state[6], state[11], state[12]);
    G(state[2], state[7], state[8], state[13]);
    G(state[3], state[4], state[9], state[14]);
}

static inline void hash_init(uint32_t state[16], const uint32_t suffix[8], uint32_t level,
                             uint32_t iteration)
{
    int i;

    memcpy(&state[0], IV, 8 * sizeof state[0]);
    state[7] ^= (level << 16) | iteration;
    for (i = 0; i < 8; i++) {
        state[8 + i] = IV[i] ^ suffix[i];
    }
}

static int hash_try(uint32_t ostate[16], uint32_t istate[16], const HashSeqSolution *proposal,
                    const Mask *mask)
{
    uint32_t f0, f1;
    int      i;

    istate[0] = IV[0] ^ proposal->s0;
    istate[1] = IV[1] ^ proposal->s1;
    memcpy(ostate, istate, 16 * sizeof ostate[0]);
    for (i = 0; i < 6; i++) {
        permute(ostate);
    }
    f0 = ostate[0];
    f1 = ostate[1];
    for (i = 2; i < 16; i += 2) {
        f0 ^= ostate[i];
        f1 ^= ostate[i + 1];
    }
    return ((f0 & mask->m0) | (f1 & mask->m1)) == 0U;
}

static void mask_from_level(Mask *mask, uint32_t level)
{
    if (level > 32U) {
        mask->m0 = ~0U;
        mask->m1 = (1U << (level - 32)) - 1U;
    } else {
        mask->m1 = 0U;
        mask->m0 = (1U << level) - 1U;
    }
}

static int solve1(HashSeqSolution *solution, uint32_t suffix[8], uint32_t level, uint32_t iteration)
{
    uint32_t        istate[16], ostate[16];
    Mask            mask;
    HashSeqSolution proposal;

    hash_init(istate, suffix, level, iteration);
    mask_from_level(&mask, level);
    proposal.s0 = proposal.s1 = 0U;
    while (hash_try(ostate, istate, &proposal, &mask) == 0) {
        if (++proposal.s0 == 0U) {
            proposal.s1++;
        }
    }
    memcpy(suffix, &ostate[8], 8 * sizeof ostate[0]);
    memcpy(solution, &proposal, sizeof *solution);

    return 0;
}

static int verify1(const HashSeqSolution *proposal, uint32_t suffix[8], uint32_t level,
                   uint32_t iteration)
{
    uint32_t state[16];
    Mask     mask;

    hash_init(state, suffix, level, iteration);
    mask_from_level(&mask, level);
    if (hash_try(state, state, proposal, &mask)) {
        memcpy(suffix, &state[8], 8 * sizeof state[0]);
        return 1;
    }
    return 0;
}

int hashseq_solve(HashSeqSolution *solutions, const uint32_t suffix[8], uint32_t level_first,
                  uint32_t level_last, uint32_t iterations)
{
    uint32_t suffix_[8];
    uint32_t level, iteration;
    int      i = 0;

    memcpy(suffix_, suffix, sizeof suffix_);
    for (level = level_first; level <= level_last; level++) {
        for (iteration = 0; iteration < iterations; iteration++) {
            if (solve1(&solutions[i++], suffix_, level, iteration) != 0) {
                return -1;
            }
        }
    }
    return 0;
}

int hashseq_verify(const HashSeqSolution *solutions, const uint32_t suffix[8], uint32_t level_first,
                   uint32_t level_last, uint32_t iterations)
{
    uint32_t suffix_[8];
    uint32_t level, iteration;
    int      i = 0;

    memcpy(suffix_, suffix, sizeof suffix_);
    for (level = level_first; level <= level_last; level++) {
        for (iteration = 0; iteration < iterations; iteration++) {
            if (verify1(&solutions[i++], suffix_, level, iteration) == 0) {
                return 0;
            }
        }
    }
    return 1;
}
